### PR TITLE
added support for project k where duplicated errors for different context are merged

### PR DESCRIPTION
**Pull request by heejaechang**
the de-duplication actually applies to linked files, shared projects and project k.

errors will merged as error for each context comes and goes. meaning, error will appear as soon as an errors comes in and refresh as more duplicates comes in.

### CODE REVIEW

**Comment by heejaechang** [comment]
@mavasani @basoundr @srivatsn @lgolding @jinujoseph @natidea can you take a look? thank you

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs [hunk 1]
```diff
@@ -4,36 +4,47 @@
...
+    /// <summary>
+    /// Base implementation of ITableDataSource
+    /// </summary>
     internal abstract class AbstractTableDataSource<TData> : ITableDataSource
     {
-        protected readonly object Gate;
-        protected readonly Dictionary<object, AbstractTableEntriesFactory<TData>> Map;
+        private readonly object _gate;
+        private readonly Dictionary<object, TableEntriesFactory<TData>> _map;
+        private readonly Dictionary<object, object> _aggregateKeyMap;
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs]: Can you please add a comment about what is stored in the key/value for this map?

Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs]: done

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs [hunk 2]
```diff
@@ -58,40 +69,138 @@ public void Shutdown()
...
+        /// </summary>
+        public abstract AbstractTableEntriesSource<TData> CreateTableEntriesSource(object data);
+
+        /// <summary>
+        /// Get unique ID for given data that will be used to find data whose items needed to be merged together.
+        /// 
+        /// for example, for linked files, data that belong to same physical file will be gathered and items that belong to
+        /// those data will be de-duplicated.
+        /// </summary>
+        protected abstract object GetOrUpdateAggregationKey(object data);
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs]: So `aggregateKeyMap` basically stores the keys for aggregate data and the value from here will be keyed on the `_map`', I see.

**Pull request merged**